### PR TITLE
Fix flaky ClientSocketTest.testSelfConnect[jvm]

### PR DESCRIPTION
**Pull request by Stexxe**
https://youtrack.jetbrains.com/issue/KTOR-2060

### CODE REVIEW

File: ktor-network/jvm/test/io/ktor/network/sockets/tests/ClientSocketTest.kt [hunk 1]
```diff
@@ -95,45 +99,35 @@ class ClientSocketTest {
...
-            val tcpSocketBuilder = aSocket(ActorSelectorManager(Dispatchers.IO)).tcp()
-            // Try to connect to that address repeatedly.
-            for (i in 0 until 100000) {
-                try {
-                    val socket = tcpSocketBuilder.connect(InetSocketAddress("127.0.0.1", port))
-                    fail("connect to self succeed: ${socket.localAddress} to ${socket.remoteAddress}")
-                } catch (ex: Exception) {
-                    // ignore
+            selector.buildOrClose({ openSocketChannel() }) {
+                val clientAddress = InetSocketAddress("0.0.0.0", 3333)
```
Comment (cy6erGn0m) [inline ktor-network/jvm/test/io/ktor/network/sockets/tests/ClientSocketTest.kt:103]: The fixed port will introduce one more instability point

File: ktor-network/jvm/test/io/ktor/network/sockets/tests/ClientSocketTest.kt [hunk 2]
```diff
@@ -95,45 +99,35 @@ class ClientSocketTest {
...
-                ServerSocket().apply {
-                    bind(InetSocketAddress("127.0.0.1", port + 1))
-                    close()
+                try {
+                    server { }
+                    nonBlocking()
+                    SocketImpl(
+                        this,
+                        clientSocket,
+                        ActorSelectorManager(Dispatchers.IO)
```
Comment (cy6erGn0m) [inline ktor-network/jvm/test/io/ktor/network/sockets/tests/ClientSocketTest.kt:121]: This should be closed explicitly. I know we had it before, but would be nice to fix once you are changing the test

File: ktor-network/jvm/test/io/ktor/network/sockets/tests/ClientSocketTest.kt [hunk 3]
```diff
@@ -95,45 +99,35 @@ class ClientSocketTest {
...
+                    ).connect(server!!.first.localSocketAddress)
+                } catch (e: ClosedChannelException) {
```
Comment (cy6erGn0m) [inline ktor-network/jvm/test/io/ktor/network/sockets/tests/ClientSocketTest.kt:123]: `e` -> `cause` or `ignored` or even `_`

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]
The main question is why we stopped looking for a random port?

**Review by Ololoshechkin — CHANGES_REQUESTED** [review]
Please, try using multiple ports but not a single one that may occasionally be taken.

**Comment by Stexxe** [comment]
We could use some mocking framework to avoid writing boilerplate code.

### Note about the fix

There are two problems with the code that fixes the self-connection problem:
1. The author assumes that the client passes a socket created from the channel and therefore closes the socket instead of the channel but it's possible to pass any socket to the `SocketImpl` constructor. If the socket isn't bound to the channel and the self-connection condition is true then the `while` loop inside `connect` method will go forever.
2. If a client encounters a self-connection problem the `ClosedChannelException` will be thrown without the actual reason why that's happened.

**Review by cy6erGn0m — APPROVED** [review]

**Review by e5l — APPROVED** [review]

**Review by Ololoshechkin — APPROVED** [review]
Please, try using multiple ports but not a single one that may occasionally be taken.

**Pull request merged**